docs/source/etcd-architecture.rst

```diff
@@ -0,0 +1,238 @@
+
+.. _etcd admin guide: https://github.com/coreos/etcd/blob/master/Documentation/admin_guide.md#optimal-cluster-size
+
+Data Storage
+~~~~~~~~~~~~
+
+etcd stores the data for the Calico network in a distributed, consistent,
+fault-tolerant manner. This set of properties ensures that the Calico network
+is always in a known-good state, while allowing for some number of the machines
+hosting etcd to fail or become unreachable.
+
+This distributed storage of Calico data also improves the ability of the Calico
+components to read from the database (which is their most common operation), as
+they can distribute their reads around the cluster.
+
+Communication
+~~~~~~~~~~~~~
+
+etcd is also used as a communication bus between components through the
+mechanism of having various components watch keys for changes. This allows the
+act of committing state to the database to cause that state to be programmed
+into the network.
+
+
+.. _calico-bgp-component:
+
+BGP Client (BIRD)
+-----------------
+
+Calico deploys a BGP client on every node that also hosts a
+:ref:`calico-felix-component`. The role of the BGP client is to read routing
+state that :ref:`calico-felix-component` programs into the kernel and
+distribute it around the data center.
+
+In Calico, this BGP component is most commonly `BIRD`_, though any BGP client
+that can draw routes from the kernel and distribute them is suitable in this
+role.
+
+The BGP client is responsible for performing the following tasks:
+
+.. _BIRD: http://bird.network.cz/
+
+Route Distribution
+~~~~~~~~~~~~~~~~~~
+
+When :ref:`calico-felix-component` inserts routes into the Linux kernel FIB,
+the BGP client will pick them up and distribute them to the other nodes in the
+deployment. This ensures that traffic is efficiently routed around the
+deployment.
+
+
+.. _calico-bgp-rr-component:
+
+BGP Route Reflector (BIRD)
+--------------------------
+
+For larger deployments, simple BGP can become a limiting factor because it
+requires every BGP client to be connected to every other BGP client in a mesh
+topology. This requires an ever increasing number of connections that rapidly
+become tricky to maintain.
+
+For that reason, in larger deployments Calico will deploy a BGP route
+reflector. This component, commonly used in the Internet, acts as a central
+point to which the BGP clients connect, preventing them from needing to talk to
+every single BGP client in the cluster.
+
+For redundancy, multiple BGP route reflectors can be deployed seamlessly.
+
+In Calico, this BGP component is also most commonly `BIRD`_, configured as a
+route reflector rather than as a standard BGP client.
+
+The BGP route reflector is responsible for the following tasks:
+
+Centralised Route Distribution
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When :ref:`calico-bgp-component` advertises routes from its FIB to the route
+reflector, the route reflector advertises those routes out to the other nodes
+in the deployment.
```

Comment:
Possibly worth mentioning here - even though it's a well known RR property - that it does this in a way that does not put itself into the routing data path.